### PR TITLE
Issue #780 Duplicate test cases when viewing an update

### PR DESCRIPTION
**Pull request by mcurlej**
Hi All, 

fixed the issue with the duplicates displaying in bodhi on test cases and assigned correct karma to them. Right now the test cases and karma is displayed correctly. Don`t really know why the dupes appeared in the DB. Current code base does not allow duplicate entries of test cases. At least this happened only for 4 packages and 22 test cases what i can tell.

### CODE REVIEW

File: bodhi/server/models.py [hunk 1]
```diff
@@ -2380,16 +2380,23 @@ def num_admin_approvals(self):
...
         return sorted(list(tests))
 
     @property
     def full_test_cases(self):
         tests = set()
         for build in self.builds:
-            for test in build.package.test_cases:
-                tests.add(test)
+            test_cases = build.package.test_cases \
+                        .order_by(TestCase.id.asc())
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: ```flake8``` is upset about the indentation on this line:

```
/home/vagrant/bodhi/bodhi/server/models.py:2394:25: E126 continuation line over-indented for hanging indent
```

File: bodhi/server/models.py [hunk 2]
```diff
@@ -2380,16 +2380,23 @@ def num_admin_approvals(self):
     def test_cases(self):
         tests = set()
         for build in self.builds:
-            for test in build.package.test_cases:
+            test_cases = build.package.test_cases \
+                        .order_by(TestCase.id.asc())
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: ```flake8``` is upset about the indentation here:

```
/home/vagrant/bodhi/bodhi/server/models.py:2384:25: E126 continuation line over-indented for hanging indent
```

Comment (bowlofeggs) [reply bodhi/server/models.py]: FWIW, the Bodhi style guide allows 100 column-wide code, so you can probably fit this on one line (I think, I didn't count ☺)

File: bodhi/server/models.py [hunk 3]
```diff
@@ -2487,6 +2494,17 @@ def url(self):
         url = '/updates/' + self.update.title + '#comment-' + str(self.id)
         return url
 
+    def unique_testcase_feedback(self):
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: Let's give this a docblock explaining what it does, and what the return type is.

File: bodhi/server/validators.py [hunk 4]
```diff
@@ -414,8 +414,8 @@ def validate_acls(request):
                 request.errors.add('body', 'builds', error_msg)
                 return
         elif acl_system == 'dummy':
-            people = (['ralph', 'bowlofeggs', 'guest'], ['guest'])
-            groups = (['ralph', 'bowlofeggs', 'guest'], ['guest'])
+            people = (['ralph', 'bowlofeggs', 'guest', 'mcurlej'], ['guest'])
+            groups = (['ralph', 'bowlofeggs', 'guest', 'mcurlej'], ['guest'])
```
Comment (bowlofeggs) [inline bodhi/server/validators.py]: It would be good to put this in its own independent commit, since it's not related to the duplicate test cases.

Also, we should probably eventually make this a setting in ```development.ini```, but I'm fine with doing it this way for now unless you want to take that on ☺

File: bodhi/tests/server/test_models.py [hunk 5]
```diff
@@ -128,6 +128,29 @@ def test_text_not_nullable(self):
         """
         self.assertEqual(model.Comment.__table__.columns['text'].nullable, False)
 
+    def test_get_unigue_testcase_feedback(self):
+        update = self.create_update(
+            (u'bodhi-2.3.3-1.fc24', u'python-fedora-atomic-composer-2016.3-1.fc24'))
+        package = update.builds[0].package
+        test1 = TestCase(name=u"Test 1", package=package)
+        test2 = TestCase(name=u"Test 2", package=package)
+        test3 = TestCase(name=u"Test 2", package=package)
```
Comment (bowlofeggs) [inline bodhi/tests/server/test_models.py]: I wonder if we should put a uniqueness constraint on name and package so there won't be any duplicates? Of course, this would mean having to write a tricky DB migration too to remove the existing duplicates…

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
There are three test failures, and these two are the non-flake8 ones:

```
Traceback (most recent call last):
  File "/home/vagrant/bodhi/bodhi/tests/server/consumers/test_masher.py", line 123, in setUp
    populate(session)
  File "/home/vagrant/bodhi/bodhi/tests/server/__init__.py", line 68, in populate
    db.flush()
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/session.py", line 2080, in flush
    self._flush(objects)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/session.py", line 2198, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/session.py", line 2162, in _flush
    flush_context.execute()
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 373, in execute
    rec.execute(self)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 485, in execute
    self.dependency_processor.process_deletes(uow, states)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/dependency.py", line 1025, in process_deletes
    secondary_update, secondary_delete)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/dependency.py", line 1096, in _run_crud
    connection = uowcommit.transaction.connection(self.mapper)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/session.py", line 294, in connection
    bind = self.session.get_bind(bindkey, **kwargs)
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/session.py", line 1315, in get_bind
    ', '.join(context)))
UnboundExecutionError: Could not locate a bind configured on mapper Mapper|Group|groups or this Session
-------------------- >> begin captured logging << --------------------
bodhi: DEBUG: Using DevBuildsys
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Test submitting a stable request to an update that already been
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/vagrant/bodhi/bodhi/tests/server/services/test_updates.py", line 2478, in test_stable_request_when_stable
    topic='update.request.stable', msg=mock.ANY)
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 937, in assert_called_with
    six.raise_from(AssertionError(_error_message(cause)), cause)
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 928, in _error_message
    msg = self._format_mock_failure_message(args, kwargs)
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 864, in _format_mock_failure_message
    actual_string = self._format_mock_call_signature(*call_args)
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 855, in _format_mock_call_signature
    return _format_call_signature(name, args, kwargs)
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 2076, in _format_call_signature
    '%s=%r' % (encode_item(key), value) for key, value in sorted(kwargs.items())
  File "/home/vagrant/bodhi/bodhi/server/models.py", line 194, in __repr__
    return '<{0} {1}>'.format(self.__class__.__name__, self.__json__())
  File "/home/vagrant/bodhi/bodhi/server/models.py", line 2445, in __json__
    for test in self.full_test_cases
  File "/home/vagrant/bodhi/bodhi/server/models.py", line 2394, in full_test_cases
    .order_by(TestCase.id.asc())
  File "<string>", line 2, in order_by
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/base.py", line 198, in generate
    self = args[0]._clone()
  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/dynamic.py", line 281, in _clone
    orm_util.instance_str(instance), self.attr.key))
DetachedInstanceError: Parent instance <RpmPackage at 0x7f2baf53b050> is not bound to a Session, and no contextual session is established; lazy load operation of attribute 'test_cases' cannot proceed
```

**Comment by ralphbean** [comment]
Doh, even after 15bdc93 there are merge conflicts.  @mcurlej, can you rebase against `develop`?

**Comment by mcurlej** [comment]
@ralphbean @bowlofeggs So all the test pass now. One test was failing because i changed the relationship in Package model. The other one was failing because a model was included twice in test_models.py.

**Comment by bowlofeggs** [comment]
@mcurlej There's another conflict ☹

File: bodhi/server/models.py [hunk 6]
```diff
@@ -2587,6 +2591,20 @@ def url(self):
...
+        """This will filter out duplicates for testcases. It will return the
+        correct number of testcases in testcase_feedbacks.
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: Let's document that it returns a list too.

File: bodhi/server/models.py [hunk 7]
```diff
@@ -2587,6 +2591,20 @@ def url(self):
...
+        """
+        feedbacks = self.testcase_feedback
+        unique_feedbacks = dict()
+        filtered_feedbacks = list()
+        for feedback in feedbacks:
+            if not unique_feedbacks.get(feedback.testcase.name):
+                unique_feedbacks[feedback.testcase.name] = True
+                filtered_feedbacks.append(feedback)
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: It seems like this is using a ```dict``` to achieve what a ```set``` is for. I recommend adding the testcase names to a set to simplify this a bit and to make it clearer what it's doing.

File: bodhi/server/models.py [hunk 8]
```diff
@@ -2587,6 +2591,20 @@ def url(self):
         url = '/updates/' + self.update.title + '#comment-' + str(self.id)
         return url
 
+    def unique_testcase_feedback(self):
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: You could make this an ```@property``` so that callers don't have to use ```()```'s.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]

**Comment by mcurlej** [comment]
@bowlofeggs using a set instead of a dict. Changed the method to a property. All tests passing.

File: bodhi/server/models.py [hunk 9]
```diff
@@ -2543,8 +2543,12 @@ def test_cases(self):
     def full_test_cases(self):
         tests = set()
         for build in self.builds:
-            for test in build.package.test_cases:
-                tests.add(test)
+            test_cases = build.package.test_cases
+            unique_tests = dict()
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: This data structure could be a ```set``` instead of a ```dict```. Since the values stored in the dict aren't important, it seems like you are mostly trying to keep a uniqueness on the keys.

File: bodhi/server/models.py [hunk 10]
```diff
@@ -2543,8 +2543,12 @@ def test_cases(self):
...
+            for test in test_cases:
+                if not unique_tests.get(test.name):
+                    unique_tests[test.name] = True
+                    tests.add(test)
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: So I think you could rework the above to be:

```
test_names = set()
for test in build.package.test_cases:
    if test.name not in test_names:
        test_names.add(test.name)
        tests.add(test)
```

In fact, it's a little curious that the ```tests``` set doesn't "uniquify" this already.

**Review by bowlofeggs — APPROVED** [review]
I think we can merge it as-is if you like, but if you want to make this one refactor I think it'd be a little cleaner. If you prefer, I can make a commit with the suggested change myself after we merge this as-is - it's up to you.

**Comment by mcurlej** [comment]
@bowlofeggs thx i have missed/forgot to re-factor also the other method. done.

**Comment by mcurlej** [comment]
@bowlofeggs thx i have missed/forgot to re-factor also the other method. done.

**Review by bowlofeggs — APPROVED** [review]
Thanks!
